### ISSUE TITLE
Empty kmer graph causes segfault

### ISSUE THREAD

**Issue by mbhall88** [rmcolq/pandora#225]
If, for some reason, a kmer graph in the index is empty, we "load" it. This causes a segmentation fault in this function https://github.com/rmcolq/pandora/blob/25d03288d6692e23c386ff29d63eaba328ce9579/src/kmergraph.cpp#L392-L419

The TODO comment in this function also hints at the fact that it isn't tested (and indeed there are no direct tests for this function)

```
    // TODO: check if tests must be updated or not due to this (I think not -
    // sorted_nodes is always sorted) if this is added, some tests bug, since it was not
    // executed before... check();
```

What an empty kmer graph does is it causes `j` (an unsigned integer) in this line https://github.com/rmcolq/pandora/blob/25d03288d6692e23c386ff29d63eaba328ce9579/src/kmergraph.cpp#L410
to have the value of -1 assigned to it. This opens a portal to that scary world in Stranger Things and all the monsters come through and cause a segmentation fault.

I have two suggestions:
1. Add tests for this function
2. Throw and exception is loading an empty kmer graph. This way the user doesn't go chasing seg faults around in a debugger for half a day

**Comment by leoisl** [comment]
> This opens a portal to that scary world in Stranger Things and all the monsters come through and cause a segmentation fault.

rofl



> I have two suggestions:
> 
> 1. Add tests for this function
> 2. Throw an exception if loading an empty kmer graph. This way the user doesn't go chasing seg faults around in a debugger for half a day

I agree with this!

Furthermore, I'd like to know the conditions in which we generate an empty kmer graph. From what I remember from the indexing code when I looked at it last year (i.e. what I will say might be wrong, just from memory), it seems that for a kmer graph to be empty, we need to have no minimizer kmers in our graph, so for some reason there is no path of size `k` or more in the PRG? In other words, we have an empty kmer graph if there exists no kmers that can be extracted from the PRG. If there is at least one kmer, then we have at least one minimizer kmer, and thus a non-empty kmer graph. I think you might have the indexing code more refreshed in your head, does this make sense?  I am just wondering if we have a bug on the code that builds a kmer graph, or if we are hitting an edge case, or if it is not that rare to build empty kmer graphs

**Comment by mbhall88** [comment]
No, the empty kmer graphs were not a pandora bug, it was a weird miscommunication between the filesystem and snakemake I think. It happened on a day when I was getting lots of zombie jobs. When I reran pandora index on the sample everything worked smoothly. So this really is an edge case.

**Comment by leoisl** [comment]
Yeah, this is tricky... I am not even sure it is even classified as edge case, as in the majority of cases when a software loads a file, it does not check for integrity in case of any filesystem issue. Should we thus add some checking for each file we load in pandora?

I don't quite understand this well... if there was any issue in the `pandora index` command, `snakemake` would recognise this and delete all the outputs, and rerun the job. But this is not the case, as it received `exit 0` from `pandora index`, so everything went fine. I think you are not also using `shadow` directive when running `pandora index`, so `snakemake` does not need to copy files from shadow dir to output dir (we could have a bug in this copy)... Could it be that when `pandora index` writes stuff to the filesystem (e.g. write a kmer graph file), we are not checking if there was any error during this writing? So maybe there was actually a filesystem error that we are not correctly capturing and erroring out.

Also regarding this, I was discussing with Zam that the index is more like a binary file, not supposed for users to view. However, the save and load functions in pandora transform it in a text format. We also have 1 kmer graph per PRG, so we also generate thousands of kmer graph files that are split in folders so that each folder has at most 4k kmer graph files... and there might be more information about the index we would like to store (like `w` and `k`). As it is a binary file, I was thinking that we could serialise the whole index object (and its kmer graphs) to the disk using for e.g. https://www.boost.org/doc/libs/1_74_0/libs/serialization/doc/index.html . I think it will be faster to read/write a single large file to disk than thousands of small files, we don't need to worry about dealing with huge number of files, whatever we add to the `Index` class will be automatically serialised and deserialised (no need to change anything), we don't need custom load/save functions (less code is better), etc... but this still falls in the code maintenance category, although it might improve performance a little bit... What are your opinions on this?

**Comment by mbhall88** [comment]
Yes, I 100% agree with making the index a single large binary file. But that is a separate (although closely-related) issue.
